STOP module

msperetto opened the pull request:
This PR is created to develop all STOP functions as a separeted module;

-STOP (open trade);
   - if enabled, create a stop order when a new trade is opened;
   - check if is a long or short order and run accordingly;

STOP (close);
   - Runs a websocket thread to listen to stream updates from the exchange;
   - If trade is closed by the stop, updates the table and notify user;
   - Must be sync with main loop to avoid double closing a trade;

msperetto committed:
Add basic configuration and constants to manage stops

Also reorganize the BASE LOCAL URL and TELEGRAM BUFFER SIZE to be more accessible throughout the code;

Refactor telegram URL and Buffer configuration to be more accessible

msperetto committed:
Add init file so the folder can be utilized as a module

Also a place to add important initialization

msperetto committed:
Add stop logic for new orders and filled orders

Also a process queue method is already created but not utilized yet;

msperetto committed:
Add new logger for the STOP module

msperetto committed:
Add binance websocket module to listen for orders filled by stops

msperetto committed:
Add method to create the stop loss order based on the order opened

msperetto committed:
Add stop service to start the threads and manage queues

Orchestrate everything necessary to manage orders filled by stops;

msperetto committed:
Add possibility to add stop orders for new created orders

msperetto committed:
Fix access to side and price from order obj

Also add a logger to se information of the order that the stop is going to be created

msperetto committed:
Add try catch when creating stop order

msperetto committed:
Add more logging and notification to better understanding of stop orders

msperetto committed:
Fix order price field to get from avgPrice and float it

msperetto committed:
Fix pair precision error when calculating price and quantity for stop

msperetto committed:
Fix Side Type comparison to get BUY instead of LONG

Fix Side Type comparison to get BUY instead of LONG

msperetto committed:
Fix stop % calculation removing /100

Also adjust round value of price

msperetto committed:
Add docker compose stop service for stop thread

msperetto committed:
Add logger for stop thread

msperetto committed:
Add more log info

msperetto committed:
Fix quantity precision calculation

msperetto committed:
Add binance logger

msperetto committed:
Fix side calculating for stop orders; has to be inverted

msperetto committed:
Fix price precision calculation for current pair

ltoledozt commented on main/marlinStop/stopLogic.py:
```diff
@@ -73,7 +74,7 @@ def create_stop_order(self, order: Dict, pair_precision) -> Dict:
             (1 - STOP_LOSS_PERCENTAGE) if order["side"] == Side_Type.LONG.value else float(order["avgPrice"]) * \
```
Fio... confesso que achei esse método do `create_stop_order` meio confuso. Acho que dá encapsular parte da lógica em novas variáveis e até num método intermediário. Por exemplo: Dá pra ter uma função só pra calcular o "stop_price". Facilita manutenção, reuso e até eventuais testes unitários...

Trabalhei com ChatGPT pra refactorarmos. Segue sugestão:

**Função separada para o cálculo do stop price:**
```
def calculate_stop_price(avg_price: float, side: str, stop_loss_percentage: float, price_precision: int) -> float:
    """
    Calcula o preço de stop baseado no lado da operação.
    Args:
        avg_price (float): Preço médio da entrada.
        side (str): 'LONG' ou 'SHORT' (Side_Type.value).
        stop_loss_percentage (float): Porcentagem de stop (ex: 0.01 para 1%).
        price_precision (int): Precisão de casas decimais.

    Returns:
        float: Preço de stop arredondado.
    """
    if side == Side_Type.LONG.value:
        stop_price = avg_price * (1 - stop_loss_percentage)
    elif side == Side_Type.SHORT.value:
        stop_price = avg_price * (1 + stop_loss_percentage)
    else:
        raise ValueError(f"Lado da operação inválido: {side}")

    return round(stop_price, price_precision)
```

**Versão refatorada do `create_stop_order`:**
```
def create_stop_order(self, order: Dict, pair_precision: int) -> Dict:
    """
    Cria uma ordem de stop baseada na ordem de entrada já executada.
    """
    marlin_stop_logger.info(f"Creating stop order for: {order}")

    try:
        avg_price = float(order["avgPrice"])
        quantity = round(float(order["origQty"]), pair_precision)
        original_side = order["side"]

        # Calcula o preço de stop
        stop_price = calculate_stop_price(
            avg_price=avg_price,
            side=original_side,
            stop_loss_percentage=STOP_LOSS_PERCENTAGE,
            price_precision=self.pair_price_precision
        )

        # Define o lado da ordem de stop (inverso da entrada)
        stop_side = Side_Type.SHORT.value if original_side == Side_Type.LONG.value else Side_Type.LONG.value

        marlin_stop_logger.info(f"Stop price: {stop_price} | Stop side: {stop_side}")

        stop_order = Binance().create_stop_loss_order(
            symbol=order["symbol"],
            quantity=quantity,
            side=stop_side,
            stop_loss_price=stop_price,
            b_id=self.api_id,
            b_sk=self.api_key
        )

        marlin_stop_logger.info(f"Stop order created: {stop_order}")
        return stop_order

    except Exception as e:
        marlin_stop_logger.error(f"Error creating stop order: {e}")
        return {"status": "error", "message": str(e)}

```

msperetto committed:
Add cancellation for all pending open orders when order is closed

msperetto committed:
Add feature to check if order was closed by stop and save in db/notify

Utilized the binance endpoint (Query account trade List); The endpoint returns all last trades for an specific symbol; We are filtering the open_time of our opened trade to make sure we only see the correct STOP order, if that happen.

Reutilizing the register_close_transaction from Negotiate() class; Just adding to the stop_order response from Binance, the missing fields of a normal closed order;

msperetto committed:
Fix positional argument position

ltoledozt commented on main/prod/tradingBot.py:
```diff
@@ -190,47 +229,85 @@ def handle_opened_trades(self):
-                continue
+            try:
+                # First check if current trade was closed by a stop order:
+                if USE_STOP_ORDERS:
```
@msperetto Acho que dá pra encapsular a lógica abaixo em métodos novos! Isso deixa o bloco mais clean, mais legível e facilita manutenção tbm!!

Sugestão:

**1. Método para checar se a ordem de stop foi fechada:**
```
def _get_closed_stop_order(self, trade):
    """
    Verifica se há uma ordem de stop fechada para o trade.
    Retorna o dicionário da ordem fechada, ou None se não houver.
    """
    stop_manager = StopManager(
        trade.pair,
        self.pairs_price_precision[trade.pair],
        self.exchange_session.e_id,
        self.exchange_session.e_sk
    )

    start_time = int(trade.open_time.timestamp() * 1000)
    closed_orders = stop_manager.check_closed_stop_order(trade.pair, start_time)
    return closed_orders[0] if closed_orders else None
```

**2. Método para tratar uma ordem de stop fechada:**
```
def _handle_closed_stop_order(self, trade, closed_order):
    """
    Registra e notifica o fechamento de uma ordem de stop.
    """
    negotiate = Negotiate(
        trade.pair,
        self.pairs_precision[trade.pair],
        self.pairs_price_precision[trade.pair],
        self.exchange_session.e_id,
        self.exchange_session.e_sk
    )

    negotiate.register_close_transaction(
        closed_order,
        trade.strategy_id,
        trade.id,
        "Stop"
    )
```

3. Uso no handle_opened_trades
```
if USE_STOP_ORDERS:
    closed_order = self._get_closed_stop_order(trade)
    if closed_order:
        self._handle_closed_stop_order(trade, closed_order)
        continue

```

ltoledozt commented on main/prod/tradingBot.py:
```diff
@@ -190,47 +229,85 @@ def handle_opened_trades(self):
+                            closed_stop_orders[0],
+                            trade.strategy_id,
+                            trade.id,
+                            "Stop"
```
Magic word??

Melhor usar: `Operation_Type.STOP` ou `Operation_Type.STOP.value` (nao tenho ctz!)

msperetto committed:
Fix response error handling for this specific endpoint:

This endpoint when running correctly, returns the dict below:
{
	"code": 200,
	"msg": "The operation of cancel all open order is done."
}

msperetto committed:
Refactor create_stop_order method to be more clear

Created local method to calculate the stop price;

Fix identation error

Fix _calculate_stop_price method by adding self

msperetto committed:
Add recvWindow to have more time to wait for binance response

ltoledozt commented on main/prod/notification.py:
```diff
@@ -0,0 +1,101 @@
+        spread,
+        profit,
+        ROI,
+        trade_type
```
Sim. Mas esse `trade_type` está sendo usado na msg do Telegram?

msperetto committed:
Refactor method handle_opened_trades(stop) to be more reusable and clean

Also using enum Operation_Type.STOP.value to substitute the string "Stop"

msperetto committed:
Fix error checking in query_account_trade_list

As it normally returns a list, we need to first check if it's a dict and then if it has "code" on it, which will indicate an error for this endpoint

msperetto committed:
Add debug message to log, to check result of closed orders

msperetto committed:
Refactor Stops % and Stop_use to come from DB

Used to get those values from config.config.py;

Solution for now is to pass the Env config obj across the whole program through classes contructors

msperetto committed:
Add debug messages for binance method query_account_trade_list

msperetto committed:
Add more debug messages to track better close stop methods

msperetto committed:
Fix input argument to Binance method qury_account_trade_list

msperetto committed:
Add debug message to check what self.pair is holding

msperetto committed:
Fix Binance requests log messages to use the use correct logger

msperetto committed:
Fix symbol passing to stop orders control

Was not passing setup parameter

msperetto committed:
Add biance debug log message

msperetto committed:
temp add test for trade list binance endpoint

msperetto committed:
Add more binance debug messages

msperetto committed:
Fix start_time passed to check stop orders

Adding 1 millisecond to make sure the current open order is not being returned

msperetto committed:
temp dropping unit test for binance endpoint

msperetto committed:
Fix instantiation of Negotiate object; missing self.setup

msperetto committed:
Add close type (stop) to telegram notification

msperetto committed:
Add unit test for binance endpoints

Added a specific logger to make easier to debug;

Added a specific logger;

msperetto committed:
Add stop (loss or gain) information to database;

Table being utilized: order_control

msperetto committed:
Update SQL scripts to get new columns and data about stops

msperetto committed:
Fix positional argument error when adding values to order table;

Adding keyword arguments after default "fees" argument so the method get the values correctly;
